File: lib/bbc/a11y/standards/exactly_one_main_heading.rb
module BBC
  module A11y
    module Standards
      class ExactlyOneMainHeading
        def initialize(page)
          @page = page
        end

        def call(errors)
          main_headings = @page.all('h1')
          count = main_headings.size
          if count == 0
            errors << "A document must have exactly one main heading." +
                      " Found 0 h1 elements."
          elsif count != 1
            errors << "A document must have exactly one main heading." +
                      " Found #{count} h1 elements:\n" +
                      main_headings.map { |h1| "* #{h1.path}" }.join("\n")
          end
        end

      end
    end
  end
end

File: lib/bbc/a11y/js/standards/exactlyOneMainHeading.js
module.exports = {
  name: 'Exactly one main heading',

  validate: function($, fail) {
    var mainHeadings = $('h1');
    var count = mainHeadings.length;
    if (count == 0) {
      fail('Found 0 h1 elements.');
    } else if (count > 1) {
      fail('Found ' + count + ' h1 elements:', mainHeadings);
    }
  }
}
